src/service-api.ts

```diff
@@ -0,0 +1,326 @@
+import extract from 'extract-zip';
+import fs from 'fs';
+import { lstat, mkdir, readdir, rm } from 'fs/promises';
+import mmm from 'mmmagic';
+import path from 'path';
+import { pipeline } from 'stream/promises';
+import util from 'util';
+import { v4 } from 'uuid';
+
+import fastifyMultipart from '@fastify/multipart';
+import { FastifyPluginAsync } from 'fastify';
+
+import { Actor, Item } from 'graasp';
+import { FileTaskManager } from 'graasp-plugin-file';
+import { ORIGINAL_FILENAME_TRUNCATE_LIMIT } from 'graasp-plugin-file-item';
+
+import {
+  H5P_ITEM_TYPE,
+  MAX_FILES,
+  MAX_FILE_SIZE,
+  MAX_NON_FILE_FIELDS,
+  TMP_EXTRACT_DIR,
+} from './constants';
+import { InvalidH5PFileError } from './errors';
+import { h5pImport } from './schemas';
+import { H5PExtra, H5PPluginOptions, PermissionLevel } from './types';
+import { H5PValidator } from './validation/h5p-validator';
+
+const magic = new mmm.Magic(mmm.MAGIC_MIME_TYPE); // don't set MAGIC_CONTINUE!
+const detectMimeType = util.promisify(magic.detectFile.bind(magic));
+
+const plugin: FastifyPluginAsync<H5PPluginOptions> = async (fastify, options) => {
+  // get services from server instance
+  const {
+    items: { taskManager: itemTaskManager },
+    itemMemberships: { taskManager: itemMembershipTaskManager },
+    taskRunner,
+  } = fastify;
+
+  const { serviceMethod, serviceOptions, pathPrefix } = options;
+
+  if (!pathPrefix) {
+    throw new Error('H5P path prefix environment variable is not defined!');
+  }
+
+  if (pathPrefix.startsWith('/')) {
+    throw new Error('H5P path prefix should not start with a "/"!');
+  }
+
+  const fileTaskManager = new FileTaskManager(serviceOptions, serviceMethod);
+  const h5pValidator = new H5PValidator();
+
+  // Helper to build the root remote path
+  const buildRootPath = (contentId: string) => path.join(pathPrefix, contentId);
+
+  // Helper to build the local or remote path of the .h5p file
+  const buildH5PPath = (rootPath: string, contentId: string) =>
+    path.join(rootPath, `${contentId}.h5p`);
+
+  // Helper to build the local or remote path of the h5p content root
+  const buildContentPath = (rootPath: string) => path.join(rootPath, 'content');
```

Comment:
Factor out in a utils file? (valid for previous functions as well)

Reply:
And next if you think so too, it might multiple utils files.

Reply:
I'm personally a big advocate of avoiding context switches, in the sense that those helpers are only used in this file, so introducing a file indirection adds cognitive load. I believe that constants that are not re-used elsewhere should also stay within the same code context. Only once you use it multiple times at multiple call sites, then should you extract them into another module for export IMHO. Though I will factor them out into an utils file, so that the structure stays coherent w.r.t. the other plugins / repos.